File: tests/test_ownership_transfer.py
from pybind11_tests import ownership_transfer as m
from pybind11_tests import ConstructorStats

import pytest
import weakref


def define_child(name, base_type, stats_type):
    # Derived instance of `DefineBase<>` in C++.
    # `stats_type` is meant to enable us to use `ConstructorStats` exclusively for a Python class.

    class ChildT(base_type):
        def __init__(self, value):
            base_type.__init__(self, value)
            self.icstats = m.get_instance_cstats(ChildT.get_cstats(), self)
            self.icstats.track_created()

        def __del__(self):
            self.icstats.track_destroyed()

        def value(self):
            # Use a different value, so that we can detect slicing.
            return 10 * base_type.value(self)

        @staticmethod
        def get_cstats():
            return ConstructorStats.get(stats_type)

    ChildT.__name__ = name
    return ChildT


ChildBad = define_child('ChildBad', m.BaseBad, m.ChildBadStats)
Child = define_child('Child', m.Base, m.ChildStats)

ChildBadUnique = define_child(
    'ChildBadUnique', m.BaseBadUnique, m.ChildBadUniqueStats)
ChildUnique = define_child(
    'ChildUnique', m.BaseUnique, m.ChildUniqueStats)


# TODO(eric.cousineau): See if this is at all possibly on PyPy.
# Placing `pytest.gc_collect` near `del` statements indicates that we are not
# capturing deletion properly.
@pytest.unsupported_on_pypy
def test_shared_ptr_derived_slicing(capture):
    from sys import getrefcount

    # [ Bad ]
    cstats = ChildBad.get_cstats()
    # Create instance in move container to permit releasing.
    obj = ChildBad(10)
    obj_weak = weakref.ref(obj)
    # This will release the reference, the refcount will drop to zero, and Python will destroy it.
    c = m.BaseBadContainer(obj)
    del obj
    # We will have lost the derived Python instance.
    assert obj_weak() is None
    # Check stats:
    assert cstats.alive() == 0
    # As an additional check, we will try to query the value from the container's value.
    # This should have been 100 if the trampoline had retained its Python portion.
    assert c.get().value() == 10
    # Destroy references.
    del c

    # [ Good ]
    # See above for setup.
    cstats = Child.get_cstats()
    # Try a temporary setup.
    c = m.BaseContainer(Child(10))
    assert cstats.alive() == 1
    obj = c.release()
    del c
    assert cstats.alive() == 1
    assert obj.value() == 100
    del obj
    assert cstats.alive() == 0
    # Use something more permanent.
    obj = Child(10)
    obj_weak = weakref.ref(obj)
    assert cstats.alive() == 1
    c = m.BaseContainer(obj)
    del obj
    assert cstats.alive() == 1
    # We now still have a reference to the object. py::wrapper<> will intercept Python's
    # attempt to destroy `obj`, is aware the `shared_ptr<>.use_count() > 1`, and will increase
    # the ref count by transferring a new reference to `py::wrapper<>` (thus reviving the object,
    # per Python's documentation of __del__).
    assert obj_weak() is not None
    assert cstats.alive() == 1
    # This goes from C++ -> Python, and then Python -> C++ once this statement has finished.
    assert c.get().value() == 100
    assert cstats.alive() == 1
    # Destroy references (effectively in C++), and ensure that we have the desired behavior.
    del c
    assert cstats.alive() == 0

    # Ensure that we can pass it from Python -> C++ -> Python, and ensure that C++ does not think
    # that it has ownership.
    obj = Child(10)
    c = m.BaseContainer(obj)
    del obj
    assert cstats.alive() == 1
    obj = c.get()
    # Now that we have it in Python, there should only be 1 Python reference, since
    # py::wrapper<> in C++ should have released its reference.
    assert getrefcount(obj) == 2
    del c
    assert cstats.alive() == 1
    assert obj.value() == 100
    del obj
    assert cstats.alive() == 0


@pytest.unsupported_on_pypy
def test_unique_ptr_derived_slicing(capture):
    # [ Bad ]
    cstats = ChildBadUnique.get_cstats()
    # Create instance in move container to permit releasing.
    try:
        c = m.BaseBadUniqueContainer(ChildBadUnique(10))
    except RuntimeError as e:
        assert "wrapper<>" in str(e)
    # We lost this portion to slicing.
    assert cstats.alive() == 0

    # [ Good ]
    cstats = ChildUnique.get_cstats()
    # Try a temporary setup.
    c = m.BaseUniqueContainer(ChildUnique(10))
    assert cstats.alive() == 1
    obj = c.release()
    del c
    assert cstats.alive() == 1
    assert obj.value() == 100
    del obj
    assert cstats.alive() == 0

    # Ensure that we can pass between Python -> C++ -> Python.
    obj = m.BaseUniqueContainer(ChildUnique(10)).release()
    obj = m.BaseUniqueContainer(obj).release()
    assert obj.value() == 100
    del obj
    assert cstats.alive() == 0

File: tests/test_ownership_transfer.cpp
/*
    tests/test_ownership_transfer.cpp -- test ownership transfer semantics.

    Copyright (c) 2017 Eric Cousineau <eric.cousineau@tri.global>

    All rights reserved. Use of this source code is governed by a
    BSD-style license that can be found in the LICENSE file.
*/

#if defined(_MSC_VER) && _MSC_VER < 1910
#  pragma warning(disable: 4702) // unreachable code in system header
#endif

#include <memory>
#include "pybind11_tests.h"
#include "object.h"

enum Label : int {
  BaseBadLabel,
  ChildBadLabel,
  BaseLabel,
  ChildLabel,

  BaseBadUniqueLabel,
  ChildBadUniqueLabel,
  BaseUniqueLabel,
  ChildUniqueLabel,
};

// For attaching instances of `ConstructorStats`.
template <int label>
class Stats {};


template <int label>
class DefineBase {
 public:
  DefineBase(int value)
      : value_(value) {
    track_created(this, value);
  }
  // clang does not like having an implicit copy constructor when the
  // class is virtual (and rightly so).
  DefineBase(const DefineBase&) = delete;
  virtual ~DefineBase() {
    track_destroyed(this);
  }
  virtual int value() const { return value_; }
 private:
  int value_{};
};

template <int label>
class DefineBaseContainer {
 public:
  using T = DefineBase<label>;
  DefineBaseContainer(std::shared_ptr<T> obj)
      : obj_(obj) {}
  std::shared_ptr<T> get() const { return obj_; }
  std::shared_ptr<T> release() { return std::move(obj_); }
 private:
  std::shared_ptr<T> obj_;
};

template <int label>
class DefineBaseUniqueContainer {
 public:
  using T = DefineBase<label>;
  DefineBaseUniqueContainer(std::unique_ptr<T> obj)
      : obj_(std::move(obj)) {}
  T* get() const { return obj_.get(); }
  std::unique_ptr<T> release() { return std::move(obj_); }
 private:
  std::unique_ptr<T> obj_;
};

template <int label>
class DefinePyBase : public DefineBase<label> {
 public:
  using BaseT = DefineBase<label>;
  using BaseT::BaseT;
  int value() const override {
    PYBIND11_OVERLOAD(int, BaseT, value);
  }
};

template <int label>
class DefinePyBaseWrapped : public py::wrapper<DefineBase<label>> {
 public:
  using BaseT = py::wrapper<DefineBase<label>>;
  using BaseT::BaseT;
  int value() const override {
    PYBIND11_OVERLOAD(int, BaseT, value);
  }
};

// BaseBad - No wrapper alias.
typedef DefineBase<BaseBadLabel> BaseBad;
typedef DefineBaseContainer<BaseBadLabel> BaseBadContainer;
typedef Stats<ChildBadLabel> ChildBadStats;

// Base - wrapper alias used in pybind definition.
typedef DefineBase<BaseLabel> Base;
typedef DefinePyBase<BaseLabel> PyBase;
typedef DefineBaseContainer<BaseLabel> BaseContainer;
typedef Stats<ChildLabel> ChildStats;

// - Unique Ptr
// BaseBad - No wrapper alias.
typedef DefineBase<BaseBadUniqueLabel> BaseBadUnique;
typedef DefineBaseUniqueContainer<BaseBadUniqueLabel> BaseBadUniqueContainer;
typedef Stats<ChildBadUniqueLabel> ChildBadUniqueStats;

// Base - wrapper alias used directly.
typedef DefineBase<BaseUniqueLabel> BaseUnique;
typedef DefinePyBaseWrapped<BaseUniqueLabel> PyBaseUnique;
typedef DefineBaseUniqueContainer<BaseUniqueLabel> BaseUniqueContainer;
typedef Stats<ChildUniqueLabel> ChildUniqueStats;

class PyInstanceStats {
 public:
  PyInstanceStats(ConstructorStats& cstats, py::handle h)
    : cstats_(cstats),
      h_(h) {}
  void track_created() {
    cstats_.created(h_.ptr());
    cstats_.value(py::str(h_).cast<std::string>());
  }
  void track_destroyed() {
    cstats_.destroyed(h_.ptr());
  }
 private:
  ConstructorStats& cstats_;
  py::handle h_;
};

PyInstanceStats get_instance_cstats(ConstructorStats& cstats, py::handle h) {
  return PyInstanceStats(cstats, h);
}

template <typename C, typename... Args>
using class_shared_ = py::class_<C, Args..., std::shared_ptr<C>>;

template <typename... Args>
using class_unique_ = py::class_<Args...>;

TEST_SUBMODULE(ownership_transfer, m) {
  // No alias - will not have lifetime extended.
  class_shared_<BaseBad>(m, "BaseBad")
      .def(py::init<int>())
      .def("value", &BaseBad::value);
  class_shared_<BaseBadContainer>(m, "BaseBadContainer")
      .def(py::init<std::shared_ptr<BaseBad>>())
      .def("get", &BaseBadContainer::get)
      .def("release", &BaseBadContainer::release);
  class_shared_<ChildBadStats>(m, "ChildBadStats");

  // Has alias - will have lifetime extended.
  class_shared_<Base, py::wrapper<PyBase>>(m, "Base")
      .def(py::init<int>())
      // Factory method for alias.
      .def(py::init([]() { return new py::wrapper<PyBase>(10); }))
      .def("value", &Base::value);
  class_shared_<BaseContainer>(m, "BaseContainer")
      .def(py::init<std::shared_ptr<Base>>())
      .def("get", &BaseContainer::get)
      .def("release", &BaseContainer::release);
  class_shared_<ChildStats>(m, "ChildStats");

  class_unique_<BaseBadUnique>(m, "BaseBadUnique")
      .def(py::init<int>())
      .def("value", &BaseBadUnique::value);
  class_unique_<BaseBadUniqueContainer>(m, "BaseBadUniqueContainer")
      .def(py::init<std::unique_ptr<BaseBadUnique>>())
      .def("get", &BaseBadUniqueContainer::get)
      .def("release", &BaseBadUniqueContainer::release);
  class_unique_<ChildBadUniqueStats>(m, "ChildBadUniqueStats");

  class_unique_<BaseUnique, PyBaseUnique>(m, "BaseUnique")
      .def(py::init<int>())
      // Factory method.
      .def(py::init([]() { return new PyBaseUnique(10); }))
      .def("value", &BaseUnique::value);
  class_unique_<BaseUniqueContainer>(m, "BaseUniqueContainer")
      .def(py::init<std::unique_ptr<BaseUnique>>())
      .def("get", &BaseUniqueContainer::get)
      .def("release", &BaseUniqueContainer::release);
  class_unique_<ChildUniqueStats>(m, "ChildUniqueStats");

  class_shared_<PyInstanceStats>(m, "InstanceStats")
      .def(py::init<ConstructorStats&, py::handle>())
      .def("track_created", &PyInstanceStats::track_created)
      .def("track_destroyed", &PyInstanceStats::track_destroyed);
  m.def("get_instance_cstats", &get_instance_cstats);
}
